source/euler_aeos/hyperbolic_system.h

```diff
@@ -1253,69 +1286,136 @@ namespace ryujin
+       */
 
-      const auto R_1 = component == 1
-                           ? vn_bar - 2. * a_bar / (gamma_bar - 1.) * x_bar
-                           : vn - 2. * a / (gamma - 1.) * x;
+      const auto R_1 = component == 1 ? vn_bar - 2. * a_bar / (gamma_bar - 1.)
+                                      : vn - 2. * a / (gamma - 1.);
 
-      /* Second Riemann characteristic: v* n + 2 / (gamma - 1) * a */
+      const auto R_2 = component == 2 ? vn_bar + 2. * a_bar / (gamma_bar - 1.)
+                                      : vn + 2. * a / (gamma - 1.);
 
-      const auto R_2 = component == 2
-                           ? vn_bar + 2. * a_bar / (gamma_bar - 1.) * x_bar
-                           : vn + 2. * a / (gamma - 1.) * x;
+      /*
+       * Note that we are really hoping for the best here... We require
+       * that R_2 >= R_1 so that we can extract a valid sound speed...
+       */
 
-      const auto s = p / ryujin::pow(rho, gamma) * ryujin::pow(x, gamma);
+      Assert(
+          R_2 >= R_1,
+          dealii::ExcMessage("Encountered R_2 < R_1 in dynamic boundary value "
+                             "enforcement. This implies that the interpolation "
+                             "with Riemann characteristics failed."));
 
       const auto vperp = m / rho - vn * normal;
 
+      const auto S = (p + pinf) * ryujin::pow(1. / rho - b, gamma);
+
+      /*
+       * Now, we have to reconstruct the actual conserved state U from the
+       * Riemann characteristics R_1, R_2, vperp, and s. We first set up
+       * {vn_new, vperp_new, a_new, S} and then solve for {rho_new, p_new}
+       * with the help of the NASG EOS surrogate formulas:
+       *
+       *   S = (p + p_infty) / rho^gamma * (1 - b * rho)^gamma
+       *
+       *   a^2 = gamma * (p + p_infty) / (rho * cov)
+       *
+       *   This implies:
+       *
+       *   a^2 / (gamma * S) = rho^{gamma - 1} / (1 - b * rho)^{1 + gamma}
+       */
+
       const auto vn_new = 0.5 * (R_1 + R_2);
+      const auto a_new_square =
+          ryujin::fixed_power<2>((gamma - 1.) / 4. * (R_2 - R_1));
+
+      /*
+       * Technically, we would need to solve for rho in the following
+       * nonlinear relationship:
+       *
+       *   a^2 / (gamma * S) = rho^{gamma - 1} / (1 - b * rho)^{gamma + 1}
+       *
+       * This seems to be a bit expensive for the fact that our dynamic
+       * boundary conditions are already terribly heuristic...
+       *
+       * So instead, we rewrite the system as:
+       *
+       *   a^2 / (gamma * S) (1 - b * rho)^2
+       *                           = (rho / (1 - b * rho))^{gamma - 1}
+       *
+       * And compute the term on the left simply with the old covolume
+       * and then solve the easier nonlinear equation on the right hand
+       * side:
+       *
+       *   A = {a^2 / (gamma * S) (1 - b * rho)^{2 gamma}}^{1/(gamma - 1)}
+       *   rho = A / (1 + b * A)
+       */
```

Comment:
@ejtovar This is slightly evil, but it avoids the nonlinear solve. It seems to perform reasonably well, at least given the fact that the whole ansatz is already questionable.

If we need something more robust we might need to think about actually implementing a whole Godunov scheme for computing boundary values…